### PR TITLE
drivers: dac: added driver for Microchip MCP4725

### PR DESCRIPTION
**Pull request by kieran-mackey**
The MCP4725 is a single channel 12 bit DAC. It is controlled via I2C.
'Write mode' bits are packed into 3 most significant bits of dac_value
parameter when using dac interface function dac_write_value().
'Write mode' controls whether the dac value is committed to nv 
eeprom memory on the MCP4725, Default write mode (no bits set) is 
'fast mode' where the 12 bit dac value is written straight to the dac 
register.

Signed-off-by: Kieran Mackey <kieran.mackey@lairdconnect.com>

### CODE REVIEW

File: drivers/dac/Kconfig.mcp4725 [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+config DAC_MCP4725
+	bool "Microchip MCP4725 DAC driver"
+	depends on I2C
+	help
+	  Enable the driver for the Microchip MCP4725.
+
+if DAC_MCP4725
+
```
Comment (galak) [inline drivers/dac/Kconfig.mcp4725:14]: seems like you can remove this if/endif block as its empty.

**Comment by martinjaeger** [comment]
I'm wondering if the BL654 DVK has the MCP4725 on the board or if you just used this board for testing it. Unfortunately, the schematic for the BL654 DVK is behind a "give-me-your-data" wall, so I couldn't figure out myself.

**Comment by kieran-mackey** [comment]
Hi Martin,
The MCP4725 was manually tested on a Laird Connectivity BL5340 which is not on zephyr master yet so this was swapped with a BL654 so CI tests could still be run. The BL654 DVK does not have an on board MCP4725.

File: drivers/dac/Kconfig.mcp4725 (hunk 1 above)
Comment (kieran-mackey) [reply drivers/dac/Kconfig.mcp4725]: Thank you, this is now fixed.

**Comment by kieran-mackey** [comment]
Actually I was mistaken on that last answer. The BL654 does have a footprint for a MCP4725 but it's not populated at the factory. BL652_DVK and BL653_DVK also have the footprint. I've added the overlay for these boards to the dac sample and test folders.

File: drivers/dac/dac_mcp4725.c [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+	bool mcp4725_ready = false;
+	int32_t retVal;
+	int32_t timeout = k_uptime_get_32() + MCP4725_BUSY_TIMEOUT_MS;
+
+	/* Wait until RDY bit is set or return error if timer exceeds MCP4725_BUSY_TIMEOUT_MS */
+	while (!mcp4725_ready) {
+		retVal = i2c_read(dev, rxData, sizeof(rxData), slave_addr);
+
+		if (retVal == 0) {
+			mcp4725_ready = (rxData[0] >> MCP4725_READ_RDY_Pos) & MCP4725_READ_RDY_Mask;
```
Comment (martinjaeger) [inline drivers/dac/dac_mcp4725.c:70]: The `MCP4725_READ_RDY_Mask` is not defined as I would usually define it. Usually the mask is applied to the register *before* shifting. In this case `MCP4725_READ_RDY_MASK` should be defined as `(1U << MCP4725_READ_RDY_POS)`, leading to the more simple statement:
```C
mcp4725_ready = rx_data[0] & MCP4725_READ_RDY_MASK;
```

File: drivers/dac/dac_mcp4725.c [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+#include <dt-bindings/dac/mcp4725.h>
+
+LOG_MODULE_REGISTER(dac_mcp4725, CONFIG_DAC_LOG_LEVEL);
+
+/* Information in this file comes from MCP4725 datasheet revision D
+ * found at https://ww1.microchip.com/downloads/en/DeviceDoc/22039d.pdf
+ */
+
+/* Bits that can't be set when an application wants to write to dac */
+#define MCP4725_VALUE_DONT_SET_BITS			0x07FFF000
```
Comment (martinjaeger) [inline drivers/dac/dac_mcp4725.c:21]: Why not?
```suggestion
#define MCP4725_VALUE_DONT_SET_BITS			0xFFFFF000
```

I would also define this as a mask for bits that are allowed, not forbidden.

File: drivers/dac/dac_mcp4725.c [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+	if (channel != 0) {
+		return -EINVAL;
+	}
+
+	/* Check value isn't over 12 bits */
+	if ((value & MCP4725_VALUE_DONT_SET_BITS) != 0) {
+		return -ENOTSUP;
+	}
+	writeMode = (value >> MCP4725_WRITE_MODE_BIT_SHIFT) & MCP4725_WRITE_MODE_Mask;
```
Comment (martinjaeger) [inline drivers/dac/dac_mcp4725.c:117]: The value passed to this function cannot contain anything else than the raw DAC value. If we really need to support both different modes, we need to think about a different solution (e.g. via devicetree).

**Review by martinjaeger — CHANGES_REQUESTED** [review]
Thanks for your contribution.

If this DAC is not present on any of the boards by default, I don't think it makes sense to configure all samples and tests for this board. I would rather suggest to take one board overlay as an example to show correct devicetree configuration. I saw that there are other breakout boards for this chip commonly available, so it could be easily adapted to other boards aswell.

I'm not completely finished with the review, but wanted to submit my comments already so that you can start addressing the style issues.

Also please squash fixing commits and force-push. See also [Contribution Workflow](https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow).

File: include/dt-bindings/dac/mcp4725.h [hunk 5]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright (c) 2021 Laird Connectivity
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_DAC_MCP4725_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_DAC_MCP4725_H_
+
+/* Config bits of MCP4725 DAC register are packed into top 5 bits of dac write value */
```
Comment (galak) [inline include/dt-bindings/dac/mcp4725.h:9]: Why are these defines here?  Meaning looking at the dts binding it doesn't seem like these defines would be used in the dts and as such these seem like they should be in a header in drivers/dac/dac_mcp4725.h or something like that.

**Review by galak — COMMENTED** [review]
Would be good to add this to `tests/drivers/build_all/`

**Review by galak — DISMISSED** [review]
Move dt-bindings/dac/mcp4725.h to drivers/dac, or explain why it needs to be in dt-bindings/

File: drivers/dac/dac_mcp4725.c [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+};
+
+struct mcp4725_data {
+	const struct device *i2c_master;
+};
+
+/* Read mcp4725 and check RDY status bit */
+static int mcp4725_wait_until_ready(const struct device *dev, uint16_t slave_addr)
+{
+	uint8_t rxData[5];
```
[the remark this answers is not in the document]
Comment (kieran-mackey) [reply drivers/dac/dac_mcp4725.c]: Change made for all relevant variables in dac_mcp4725.c

File: drivers/dac/dac_mcp4725.c [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+/* Information in this file comes from MCP4725 datasheet revision D
+ * found at https://ww1.microchip.com/downloads/en/DeviceDoc/22039d.pdf
+ */
+
+/* Bits that can't be set when an application wants to write to dac */
+#define MCP4725_VALUE_DONT_SET_BITS			0x07FFF000
+
+/* Defines for field values in MCP4725 DAC register */
+#define MCP4725_WRITE_MODE_Mask				0x7
```
[the remark this answers is not in the document]
Comment (kieran-mackey) [reply drivers/dac/dac_mcp4725.c]: Change made for _Mask and _Pos defines

File: drivers/dac/dac_mcp4725.c (hunk 2 above)
Comment (kieran-mackey) [reply drivers/dac/dac_mcp4725.c]: Change made

File: drivers/dac/dac_mcp4725.c (hunk 3 above)
Comment (kieran-mackey) [reply drivers/dac/dac_mcp4725.c]: This #define has been replaced with a #define for max value to check the input isn't over 12 bits

File: drivers/dac/dac_mcp4725.c [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+#define MCP4725_POWER_DOWN_Mask				0x3
+#define MCP4725_DAC_VAL_Mask                            0xFFF
+
+#define MCP4725_FAST_MODE_POWER_DOWN_Pos		4
```
[the remark this answers is not in the document]
Comment (kieran-mackey) [reply drivers/dac/dac_mcp4725.c]: Change made

File: drivers/dac/dac_mcp4725.c (hunk 4 above)
Comment (kieran-mackey) [reply drivers/dac/dac_mcp4725.c]: I've removed the configuration bits and with it the "WRITE_TO_EEPROM" mode. Only extended functionality was removed so this dac still functions with the standard interface. I'm open to adding this functionality back in a more suitable way in a later pull request. I may need some hints on what to do there - if it's added to the device tree does that mean the extended configuration is set at compile time? Is there a way to add optional functions to the dac_driver_api interface?

File: include/dt-bindings/dac/mcp4725.h (hunk 5 above)
Comment (kieran-mackey) [reply include/dt-bindings/dac/mcp4725.h]: You're right, they should have belonged in drivers/dac but the whole header file has been removed in the last commit as the defines are no longer used in code.

File: drivers/dac/dac_mcp4725.c [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+#define MCP4725_BUSY_TIMEOUT_MS			60U
+
+struct mcp4725_config {
+	const char *i2c_master_dev_name;
+	uint16_t i2c_slave_addr;
+};
+
+struct mcp4725_data {
+	const struct device *i2c_master;
+};
```
Comment (martinjaeger) [inline drivers/dac/dac_mcp4725.c:42]: The terms master & slave [should not be used](https://docs.zephyrproject.org/latest/guides/coding_guidelines/index.html#rule-a-2-inclusive-language) anymore. In addition to that I suggest to switch to new (more efficient) `DEVICE_DT_GET` API to obtain the device (see below). The `_data` struct is not needed anymore afterwards.
```suggestion
struct mcp4725_config {
	const struct device *i2c_dev;
	uint16_t i2c_addr;
};
```

File: drivers/dac/dac_mcp4725.c [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+static int dac_mcp4725_init(const struct device *dev)
+{
+	struct mcp4725_data *data = dev->data;
+	const struct mcp4725_config *config = dev->config;
+
+	data->i2c_master = device_get_binding(
+		config->i2c_master_dev_name);
+	if (!data->i2c_master) {
```
Comment (martinjaeger) [inline drivers/dac/dac_mcp4725.c:124]:
```suggestion
	if (!device_is_ready(config->i2c_dev)) {
```

File: drivers/dac/dac_mcp4725.c [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+		.i2c_master_dev_name = DT_INST_BUS_LABEL(index),	\
+		.i2c_slave_addr = DT_INST_REG_ADDR(index)		\
+		};							\
+									\
+	DEVICE_DT_INST_DEFINE(index, dac_mcp4725_init,			\
+			    device_pm_control_nop,			\
+			    &mcp4725_data_##index,			\
+			    &mcp4725_config_##index, POST_KERNEL,	\
+			    CONFIG_DAC_MCP4725_INIT_PRIORITY,		\
+			    &mcp4725_driver_api);
```
Comment (martinjaeger) [inline drivers/dac/dac_mcp4725.c:156]:
```suggestion
#define INST_DT_MCP4725(index)						\
	static const struct mcp4725_config mcp4725_config_##index = {	\
		.i2c_dev = DEVICE_DT_GET(DT_INST_BUS(index)),	\
		.i2c_addr = DT_INST_REG_ADDR(index)		\
	};								\
									\
	DEVICE_DT_INST_DEFINE(index, dac_mcp4725_init,			\
			    device_pm_control_nop,			\
			    NULL,			\
			    &mcp4725_config_##index, POST_KERNEL,	\
			    CONFIG_DAC_MCP4725_INIT_PRIORITY,		\
			    &mcp4725_driver_api);
```

**Review by martinjaeger — CHANGES_REQUESTED** [review]
Can you please also add remarks to the board, sample and test documentation (e.g. Readme files) that this DAC is not populated on the boards by default? Otherwise people might get confused, as the DAC option is not even mentioned in manuals like [this one](http://assets.lairdtech.com/home/brandworld/files/BL652%20Dev%20Kit%20User%20Guide.pdf).

After that we should be ready to go from my side.

**Comment by kieran-mackey** [comment]
Thank you for the comments @martinjaeger, I've added notes about the MCP4725 not being factory populated on BL652/3/4 to the sample readme and test dts files where appropriate.

File: drivers/dac/dac_mcp4725.c (hunk 9 above)
Comment (kieran-mackey) [reply drivers/dac/dac_mcp4725.c]: Change made

File: drivers/dac/dac_mcp4725.c (hunk 11 above)
Comment (kieran-mackey) [reply drivers/dac/dac_mcp4725.c]: Change made

File: drivers/dac/dac_mcp4725.c (hunk 10 above)
Comment (kieran-mackey) [reply drivers/dac/dac_mcp4725.c]: Change made

File: drivers/dac/dac_mcp4725.c [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+	return ret;
+}
+
+static int dac_mcp4725_init(const struct device *dev)
+{
+	const struct mcp4725_config *config = dev->config;
+
+	if (!device_is_ready(config->i2c_dev)) {
+		LOG_ERR("I2C device not found: %s",
+			    config->i2c_dev);
```
Comment (martinjaeger) [inline drivers/dac/dac_mcp4725.c:118]: This is not a string anymore, so we should probably simply use:
```suggestion
		LOG_ERR("I2C device not found");
```

File: drivers/dac/dac_mcp4725.c [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+	.channel_setup = mcp4725_channel_setup,
+	.write_value = mcp4725_write_value,
+};
+
+
+#define INST_DT_MCP4725(index)						\
+	static const struct mcp4725_config mcp4725_config_##index = {	\
+		.i2c_dev = DEVICE_DT_GET(DT_INST_BUS(index)),		\
+		.i2c_addr = DT_INST_REG_ADDR(index)			\
+		};							\
```
Comment (martinjaeger) [inline drivers/dac/dac_mcp4725.c:140]: Nits: Wrong indentation.
```suggestion
	};								\
```

**Review by martinjaeger — CHANGES_REQUESTED** [review]

File: drivers/dac/dac_mcp4725.c (hunk 12 above)
Comment (kieran-mackey) [reply drivers/dac/dac_mcp4725.c]: Change made

File: drivers/dac/dac_mcp4725.c (hunk 13 above)
Comment (kieran-mackey) [reply drivers/dac/dac_mcp4725.c]: Change made

**Review by martinjaeger — APPROVED** [review]
Thanks, looks good to me now.

**Comment by martinjaeger** [comment]
@carlescufi did you re-request my review on purpose or by accident? (I approved already)

File: samples/drivers/dac/sample.yaml [hunk 14]
```diff
@@ -5,7 +5,8 @@ tests:
     tags: DAC
     platform_allow: |
       arduino_zero frdm_k22f frdm_k64f nucleo_f091rc nucleo_g071rb nucleo_g431rb
-      nucleo_l073rz nucleo_l152re twr_ke18f nucleo_f767zi nucleo_f429zi
+      nucleo_l073rz nucleo_l152re twr_ke18f nucleo_f767zi nucleo_f429zi bl652_dvk
+      bl653_dvk bl654_dvk
```
Comment (MaureenHelm) [inline samples/drivers/dac/sample.yaml:9]: This is beyond the scope of this PR, but @martinjaeger can you look at removing this list and just using the `depends_on: dac` below?

**Review by MaureenHelm — APPROVED** [review]

**Comment by MaureenHelm** [comment]
@galak please revisit

File: samples/drivers/dac/sample.yaml (hunk 14 above)
Comment (martinjaeger) [reply samples/drivers/dac/sample.yaml:9]: Good point. Unfortunately, the supported resolution and channel number of the DAC cannot be obtained from devicetree, so we still have some `#ifdef`s in the sample code to determine it based on the selected board. I'll have a look if we can switch the `#ifdef`s to SOC-based versions and get rid of this growing allow-list. But that should not block this PR.

**Review by martinjaeger — APPROVED** [review]

**Comment by kieran-mackey** [comment]
Hi @galak, the file mentioned in your previous comment has since been deleted. Are there any further comments you have for this pull request?

**Comment by carlescufi** [comment]
> @carlescufi did you re-request my review on purpose or by accident? (I approved already)

Probably an oversight, sorry.

**Pull request merged**